File: packages/skills/using-openai-platform/templates/chat-completion.template.ts
/**
 * OpenAI Chat Completions Template (TypeScript)
 *
 * This template provides a type-safe chat completion implementation
 * with proper error handling.
 *
 * Placeholders:
 * - {{model}} - Model ID (e.g., "gpt-5")
 * - {{system_prompt}} - System message content
 *
 * Usage:
 *   npx ts-node chat-completion.ts "What is TypeScript?"
 */

import OpenAI from 'openai';
import type { ChatCompletionMessageParam } from 'openai/resources/chat/completions';

// Configuration
const MODEL = '{{model}}';
const SYSTEM_PROMPT = '{{system_prompt}}';
const MAX_TOKENS = 1000;
const TEMPERATURE = 0.7;

/**
 * Chat response with usage information
 */
interface ChatResponse {
  content: string;
  usage: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
  };
}

/**
 * OpenAI Chat Completions client with TypeScript typing
 */
class ChatClient {
  private client: OpenAI;
  private model: string;
  private systemPrompt: string;
  private maxTokens: number;
  private temperature: number;
  private messages: ChatCompletionMessageParam[];

  constructor(options: {
    model?: string;
    systemPrompt?: string;
    maxTokens?: number;
    temperature?: number;
  } = {}) {
    this.client = new OpenAI();
    this.model = options.model ?? MODEL;
    this.systemPrompt = options.systemPrompt ?? SYSTEM_PROMPT;
    this.maxTokens = options.maxTokens ?? MAX_TOKENS;
    this.temperature = options.temperature ?? TEMPERATURE;
    this.messages = [];

    // Initialize with system prompt
    if (this.systemPrompt) {
      this.messages.push({
        role: 'system',
        content: this.systemPrompt
      });
    }
  }

  /**
   * Send a message and get a response
   */
  async chat(userMessage: string): Promise<ChatResponse> {
    // Add user message to history
    this.messages.push({
      role: 'user',
      content: userMessage
    });

    try {
      console.log(`Sending request to ${this.model}`);

      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: this.messages,
        max_tokens: this.maxTokens,
        temperature: this.temperature
      });

      const assistantMessage = response.choices[0].message.content ?? '';

      // Add to history
      this.messages.push({
        role: 'assistant',
        content: assistantMessage
      });

      // Log usage
      console.log(
        `Tokens used - Prompt: ${response.usage?.prompt_tokens}, ` +
        `Completion: ${response.usage?.completion_tokens}`
      );

      return {
        content: assistantMessage,
        usage: {
          promptTokens: response.usage?.prompt_tokens ?? 0,
          completionTokens: response.usage?.completion_tokens ?? 0,
          totalTokens: response.usage?.total_tokens ?? 0
        }
      };

    } catch (error) {
      if (error instanceof OpenAI.AuthenticationError) {
        console.error('Authentication failed - check your API key');
        throw error;
      }

      if (error instanceof OpenAI.RateLimitError) {
        console.error('Rate limit exceeded - implement backoff');
        throw error;
      }

      if (error instanceof OpenAI.APIError) {
        console.error(`API error: ${error.message}`);
        throw error;
      }

      throw error;
    }
  }

  /**
   * Clear conversation history, keeping system prompt
   */
  clearHistory(): void {
    this.messages = [];
    if (this.systemPrompt) {
      this.messages.push({
        role: 'system',
        content: this.systemPrompt
      });
    }
  }

  /**
   * Get the current conversation history
   */
  getHistory(): ChatCompletionMessageParam[] {
    return [...this.messages];
  }
}

/**
 * Main entry point
 */
async function main(): Promise<void> {
  const userInput = process.argv.slice(2).join(' ');

  if (!userInput) {
    console.log('Usage: npx ts-node chat-completion.ts <message>');
    process.exit(1);
  }

  try {
    const client = new ChatClient();
    const response = await client.chat(userInput);
    console.log(response.content);

  } catch (error) {
    if (error instanceof OpenAI.AuthenticationError) {
      console.error('Error: Invalid API key. Set OPENAI_API_KEY environment variable.');
      process.exit(1);
    }

    if (error instanceof OpenAI.RateLimitError) {
      console.error('Error: Rate limit exceeded. Please try again later.');
      process.exit(1);
    }

    console.error(`Error: ${error}`);
    process.exit(1);
  }
}

main();

export { ChatClient, ChatResponse };

File: packages/skills/using-openai-platform/templates/chat-completion.template.py
#!/usr/bin/env python3
"""
OpenAI Chat Completions Template

This template provides a production-ready chat completion implementation
with proper error handling and configuration.

Placeholders:
- {{model}} - Model ID (e.g., "gpt-5")
- {{system_prompt}} - System message content
- {{max_tokens}} - Maximum response tokens

Usage:
    python chat-completion.py "What is Python?"
"""

import os
import sys
import logging
from typing import List, Dict, Optional

from openai import OpenAI, APIError, RateLimitError, AuthenticationError

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Configuration
MODEL = "{{model}}"
SYSTEM_PROMPT = "{{system_prompt}}"
MAX_TOKENS = {{max_tokens}}
TEMPERATURE = 0.7


class ChatClient:
    """OpenAI Chat Completions client with error handling."""

    def __init__(
        self,
        model: str = MODEL,
        system_prompt: str = SYSTEM_PROMPT,
        max_tokens: int = MAX_TOKENS,
        temperature: float = TEMPERATURE
    ):
        """Initialize the chat client.

        Args:
            model: OpenAI model ID
            system_prompt: System message for the conversation
            max_tokens: Maximum tokens in response
            temperature: Sampling temperature (0-2)
        """
        self.client = OpenAI()
        self.model = model
        self.system_prompt = system_prompt
        self.max_tokens = max_tokens
        self.temperature = temperature
        self.messages: List[Dict[str, str]] = []

        # Initialize with system prompt
        if self.system_prompt:
            self.messages.append({
                "role": "system",
                "content": self.system_prompt
            })

    def chat(self, user_message: str) -> str:
        """Send a message and get a response.

        Args:
            user_message: The user's message

        Returns:
            The assistant's response

        Raises:
            AuthenticationError: Invalid API key
            RateLimitError: Rate limit exceeded
            APIError: Other API errors
        """
        # Add user message to history
        self.messages.append({
            "role": "user",
            "content": user_message
        })

        try:
            logger.info(f"Sending request to {self.model}")

            response = self.client.chat.completions.create(
                model=self.model,
                messages=self.messages,
                max_tokens=self.max_tokens,
                temperature=self.temperature
            )

            # Extract response content
            assistant_message = response.choices[0].message.content

            # Add to history
            self.messages.append({
                "role": "assistant",
                "content": assistant_message
            })

            # Log usage
            logger.info(
                f"Tokens used - Prompt: {response.usage.prompt_tokens}, "
                f"Completion: {response.usage.completion_tokens}"
            )

            return assistant_message

        except AuthenticationError as e:
            logger.error("Authentication failed - check your API key")
            raise

        except RateLimitError as e:
            logger.error("Rate limit exceeded - implement backoff")
            raise

        except APIError as e:
            logger.error(f"API error: {e}")
            raise

    def clear_history(self) -> None:
        """Clear conversation history, keeping system prompt."""
        self.messages = []
        if self.system_prompt:
            self.messages.append({
                "role": "system",
                "content": self.system_prompt
            })

    def get_history(self) -> List[Dict[str, str]]:
        """Get the current conversation history."""
        return self.messages.copy()


def main():
    """Main entry point for CLI usage."""
    if len(sys.argv) < 2:
        print("Usage: python chat-completion.py <message>")
        sys.exit(1)

    user_input = " ".join(sys.argv[1:])

    try:
        client = ChatClient()
        response = client.chat(user_input)
        print(response)

    except AuthenticationError:
        print("Error: Invalid API key. Set OPENAI_API_KEY environment variable.")
        sys.exit(1)

    except RateLimitError:
        print("Error: Rate limit exceeded. Please try again later.")
        sys.exit(1)

    except APIError as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
